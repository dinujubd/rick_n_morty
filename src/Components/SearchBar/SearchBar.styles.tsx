import { createStyles, makeStyles, Theme } from "@material-ui/core";

export const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            width: '90%',
            margin: theme.spacing(1),
            display: 'flex'
        },
        input: {
            marginLeft: theme.spacing(1),
            flex: 1,
        },
        iconButton: {
            padding: 10,
        }
    }),
);
