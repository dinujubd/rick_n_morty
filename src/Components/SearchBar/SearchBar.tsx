import React from "react";
import { createStyles, debounce, IconButton, InputBase, makeStyles, Modal, Paper, Theme, Box, Grid, Select, MenuItem, Button, NativeSelect } from "@material-ui/core";
import MenuIcon from '@material-ui/icons/Menu';
import SearchIcon from '@material-ui/icons/Search';
import { connect } from "react-redux"
import { searchAction } from '../../store/actions'
import { ApplicationDispatch } from "../../store/store";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            maxWidth: 345,
            margin: '5px',
            display: 'flex'
        },
        input: {
            marginLeft: theme.spacing(1),
            flex: 1,
        },
        iconButton: {
            padding: 10,
        },
        paper: {
            position: 'absolute',
            width: 400,
            backgroundColor: theme.palette.background.paper,
            boxShadow: theme.shadows[5],
            padding: theme.spacing(2, 4, 3),
            top: '50%',
            left: '50%',
            transform: 'translate(-50%, -50%)'
        },
        fiterSelect: {
            width: '100%'
        },
    }),
);

interface DispatchProps {
    search: (searchConfig: SearchConfig) => void
}

interface SearchConfig {
    name?: string,
    status?: string,
    species?: string,
    type?: string,
    gender?: string,
}

const SearchBar: React.FC<DispatchProps> = (props) => {

    const classes = useStyles();
    const [open, setOpen] = React.useState(false);
    const [searchConfig, setSearchConfig] = React.useState<SearchConfig>({});

    const changeHandler = (event: any) => {
        setSearchConfig({...searchConfig, name: event.target.value})
        props.search({ name: event.target.value })
    };

    const debouncedChangeHandler = React.useCallback(
        debounce(changeHandler, 300)
        , []);

    const handleClose = React.useCallback(() => {
        setOpen(false)
    }, [])


    const changeSearchFilterHandler = (key: keyof SearchConfig) => (e: any) => {
        const value = e.nativeEvent.target.value;
        if (value && value === '*')
            setSearchConfig({ ...searchConfig, [key]: null });
        else
            setSearchConfig({ ...searchConfig, [key]: value });
    }

    const handleSearchFilter = React.useCallback(() => {
        props.search(searchConfig);
        setOpen(false);
    }, [searchConfig])

    return (
        <Paper component="form" className={classes.root}>
            <Modal
                open={open}
                onClose={handleClose}
                aria-labelledby="simple-modal-title"
                aria-describedby="simple-modal-description"
            >
                <div className={classes.paper}>
                    <h2 >Search Filter</h2>
                    <Grid
                        container
                        justify="flex-start"
                    >
                        <Grid xs={6}>Status</Grid>
                        <Grid xs={6}>
                            <NativeSelect onChange={changeSearchFilterHandler('status')} value={searchConfig.status || '*'} className={classes.fiterSelect}>
                                <option value="*">Any</option>
                                <option value="Alive">Alive</option>
                                <option value="Dead">Dead</option>
                                <option value="unknown">unknown</option>
                            </NativeSelect>
                        </Grid>
                        <Grid xs={6}>Species</Grid>
                        <Grid xs={6}>
                            <Select onChange={changeSearchFilterHandler('species')} value={searchConfig.species || '*'} className={classes.fiterSelect}>
                                <option value="*">Any</option>
                                <option value="Human">Human</option>
                                <option value="Alien">Alien</option>
                                <option value="Humanoid">Humanoid</option>
                                <option value="Mythological Creature">Mythological Creature</option>
                                <option value="Poopybutthole">Poopybutthole</option>
                                <option value="Robot">Robot</option>
                                <option value="Cronenberg">Cronenberg</option>
                                <option value="Disease">Disease</option>
                                <option value="unknown">unknown</option>
                            </Select>
                        </Grid>
                        <Grid xs={6}>Type</Grid>
                        <Grid xs={6}>
                            <Select onChange={changeSearchFilterHandler('type')} value={searchConfig.type || '*'} className={classes.fiterSelect}>
                                <option value="*">Any</option>
                                <option value="Genetic experiment">Genetic experiment</option>
                                <option value="Superhuman (Ghost trains summoner)">Superhuman (Ghost trains summoner)</option>
                                <option value="Parasite">Parasite</option>
                                <option value="Human with antennae">Human with antennae</option>
                                <option value="Human with ants in his eyes">Human with ants in his eyes</option>
                            </Select>
                        </Grid>
                        <Grid xs={6}>Gender</Grid>
                        <Grid xs={6}>
                            <Select onChange={changeSearchFilterHandler('gender')} value={searchConfig.gender || '*'} className={classes.fiterSelect}>
                                <option value="*">Any</option>
                                <option value="Male">Male</option>
                                <option value="Female">Female</option>
                                <option value="unknown">Unknown</option>

                            </Select>


                        </Grid>
                        <Button onClick={handleSearchFilter} variant="contained" color="secondary">Search</Button>
                    </Grid>
                </div>
            </Modal>
            <IconButton className={classes.iconButton} aria-label="menu">
                <MenuIcon onClick={() => setOpen(true)} />
            </IconButton>
            <InputBase
                className={classes.input}
                placeholder="Search Character"
                inputProps={{ 'aria-label': 'Search Character' }}
                onChange={debouncedChangeHandler}
            />
            <IconButton type="submit" className={classes.iconButton} aria-label="search">
                <SearchIcon />
            </IconButton>
        </Paper>

    )
}

const mapDispatchToProps = (dispatch: ApplicationDispatch): DispatchProps => {
    return {
        search: (searchConfig: SearchConfig) => dispatch(searchAction(searchConfig))
    }
}

export default connect(_ => _, mapDispatchToProps)(SearchBar)