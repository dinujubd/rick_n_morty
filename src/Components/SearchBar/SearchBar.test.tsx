import * as React from 'react';
import renderer from 'react-test-renderer';
import { SearchBar } from './SearchBar';

describe('<SearchBar />', () => {
    it('should renders correctly', () => {
        const tree = renderer
            .create(<SearchBar search={jest.fn()}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});