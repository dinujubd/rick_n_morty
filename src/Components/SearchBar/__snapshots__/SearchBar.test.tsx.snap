// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> should renders correctly 1`] = `
<form
  className="MuiPaper-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
>
  <button
    aria-label="menu"
    className="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-3"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
    </span>
  </button>
  <div
    className="MuiInputBase-root makeStyles-input-2"
    onClick={[Function]}
  >
    <input
      aria-label="Search Character"
      className="MuiInputBase-input"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Search Character"
      type="text"
    />
  </div>
  <button
    aria-label="search"
    className="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-3"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </span>
  </button>
</form>
`;
